### PR TITLE
Add Observation implementation

### PR DESCRIPTION
**Pull request by johkin**
Initial implementation of observation for spring-ws.

Please feel free to comment and provide feedback.

#1094

### CODE REVIEW

**Comment by johkin** [comment]
Ping @corneil

**Comment by corneil** [comment]
@johkin Thanks for this great contribution. 
I would suggest adding something like: `Assert.notNull(observation, "Expected observation in messageContext");` at appropriate places.

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java
Comment (marcingrzejszczak) [inline spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java:34]: Can any of these values be null? If that's the case you must not set the key value with a null

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java
Comment (marcingrzejszczak) [inline spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java:101]: Contextual name could be more about the action that we're performing. For http that's e.g. http post /foo

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceTemplateObservationDocumentation.java
Comment (marcingrzejszczak) [inline spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceTemplateObservationDocumentation.java:44]: You can remove this

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceTemplateObservationDocumentation.java
Comment (marcingrzejszczak) [inline spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceTemplateObservationDocumentation.java:79]: Javadocs missing

File: spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java
Comment (marcingrzejszczak) [inline spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java:35]: Values can be null?

File: spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java
Comment (marcingrzejszczak) [inline spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java:95]: Same remark as with the previous one

File: spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/EndpointObservationDocumentation.java
Comment (marcingrzejszczak) [inline spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/EndpointObservationDocumentation.java:31]: Javadocs missing

File: spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/EndpointObservationDocumentation.java
Comment (marcingrzejszczak) [inline spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/EndpointObservationDocumentation.java:44]: You can remove this

File: spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/EndpointObservationDocumentation.java
Comment (marcingrzejszczak) [inline spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/EndpointObservationDocumentation.java:71]: Javadocs missing

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java
Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java:35]: All of the values are either "unknown" or "none" if no value is present.

File: spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java
Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java:35]: All of the keys are either "unknown" or "none" if no value is present.

File: spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java
Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java]: I'm not sure how this value is used? What would be a good implementation in your opinion? This implementation describes the context as a WebServiceEndpoint that handles a payload with the given namespace and local-part.

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java
Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java]: I'm not sure how this value is used? What would be a good implementation in your opinion?

My new implementation is perhaps a bit too complicated, it will return "WebService host:port/path SoapAction=Hello QName=http://mynamespace.com:rootElement"

Comment (marcingrzejszczak) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java]: If we follow the OTel semantic convention suggestions (https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#name) we could go with sth like `POST /path` ? Then the rest of the information like soapaction, qname etc. could go as tags. WDYT?

Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/DefaultWebServiceTemplateConvention.java]: `Post /path` sounds reasonable.

File: spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java
Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/DefaultWebServiceEndpointConvention.java]: After reading https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#http-client-server-example, would it be ok to use the same ContextualName for a WebService Endpoint?
`POST /path`

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java
Comment (shakuzen) [inline spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java:57]: Nit pick: this and other fields can be final.

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java
Comment (shakuzen) [inline spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java:145]: I'm curious: why make a setter for this instead of take the custom convention in the constructor? It allows for users to change the custom convention at runtime, but is that something you expect users will want to do? I would err on the side of making it final and taking it in the constructor until a user requests being able to set it after construction and provides more context around why they want to do that.

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java
Comment (shakuzen) [inline spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java:102]: What's the effect of throwing an exception here? Will it result in an exception thrown from the call to WebServiceTemplate? I would generally advise against doing that. If the instrumentation is broken, we do want some way to notice that (logging perhaps), but user code that works without observability instrumentation should not fail when observability instrumentation is added. An exception to this rule of thumb might be for configuration issues where an exception on startup can be thrown, but this will be thrown when handling a request, if I understand correctly.

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java
Comment (shakuzen) [inline spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java:134]: Could this path be high cardinality? I suppose by convention that isn't usually expected with SOAP endpoints, but if this might contain a path variable, we may want to consider a way to resolve a templated path instead of the raw path.

**Review by shakuzen — COMMENTED** [review]
Thank you for the work on this. It's a very complete pull request. I left some feedback for consideration.

File: spring-ws-core/src/test/java/org/springframework/ws/client/core/observation/WebServiceTemplateObservationIntegrationTest.java
Comment (johkin) [reply spring-ws-core/src/test/java/org/springframework/ws/client/core/observation/WebServiceTemplateObservationIntegrationTest.java:302]: There should definitely be assertions in a test, I was a bit quick to push the PR before revisiting the tests!

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java
Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java:102]: I agree with your remark, the throws-statement is just a leftover from the generated method and should of course be removed if it isn't used.

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java
Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java:134]: Good point!
As far as my experience goes, I've never seen a WebService published with a variable path but it could certainly be the case.
The quick fix could then be to add it to the HighCardinalityKeys, I'm not sure how to hold of a templated path.

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceTemplateObservationContext.java
Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceTemplateObservationContext.java]: I added the WarnThenDebugLogger, seems like a reasonable solution.

File: spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/WebServiceEndpointContext.java
Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/server/endpoint/observation/WebServiceEndpointContext.java]: Replaced with WarnThenDebugLogger

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java
Comment (shakuzen) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java:102]: I meant the call to `Assert.nonNull` will throw an exception if `observation` is null, and we probably don't want that.

Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java:102]: Could this be a candidate for WarnThenDebugLogger, or should it just silently be ignored?

Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java]: Replaced the assertion with WarnThenDebugLogger

File: spring-ws-core/src/test/java/org/springframework/ws/server/endpoint/observation/ObservationInterceptorIntegrationTest.java
Comment (shakuzen) [inline spring-ws-core/src/test/java/org/springframework/ws/server/endpoint/observation/ObservationInterceptorIntegrationTest.java:109]: What is the `path` key-value in this test case?

Comment (johkin) [reply spring-ws-core/src/test/java/org/springframework/ws/server/endpoint/observation/ObservationInterceptorIntegrationTest.java:109]: If the current request path fully matches the registered servlet path, there will be no varying high cardinality key. Would it be better to add a 'none'-value?

**Comment by johkin** [comment]
I did some minor restructuring of the code to improve readability and error-handling/logging.

The method used to find the root element is identical in both interceptors, could be and improvement to refactor this to a helper-class? On the other hand, this requires the helper to be injected in the interceptors and creates another dependency.

File: spring-ws-core/src/test/java/org/springframework/ws/server/endpoint/observation/ObservationInterceptorIntegrationTest.java
Comment (johkin) [reply spring-ws-core/src/test/java/org/springframework/ws/server/endpoint/observation/ObservationInterceptorIntegrationTest.java:110]: I think I misunderstood your comment, I have now added the path-key to the observation. The high cardinality key 'pathInfo' is only provided when the request has a wildcard match.

Comment (shakuzen) [reply spring-ws-core/src/test/java/org/springframework/ws/server/endpoint/observation/ObservationInterceptorIntegrationTest.java:110]: My comment could have been more clear. I was expecting if there is a `path` key-value that it would be in the assertions here, but it isn't.

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/RootElementSAXHandler.java
Comment (shakuzen) [inline spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/RootElementSAXHandler.java]: Should this be public or can it be package-private?

File: spring-ws-core/src/test/java/org/springframework/ws/server/endpoint/observation/ObservationInterceptorIntegrationTest.java
Comment (johkin) [reply spring-ws-core/src/test/java/org/springframework/ws/server/endpoint/observation/ObservationInterceptorIntegrationTest.java:110]: I added the path key to the test, but it doesn't seem to show in this view?

**Review by shakuzen — APPROVED** [review]

File: spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/RootElementSAXHandler.java
Comment (shakuzen) [reply spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/RootElementSAXHandler.java]: Oops. I was looking at old commits.

**Review by marcingrzejszczak — APPROVED** [review]

File: spring-ws-core/src/main/java/org/springframework/ws/support/ObservationHelper.java
Comment (shakuzen) [inline spring-ws-core/src/main/java/org/springframework/ws/support/ObservationHelper.java:52]: Is there a reason to having multiple instances of this class? Could it be a singleton utility class instead?

Comment (shakuzen) [reply spring-ws-core/src/main/java/org/springframework/ws/support/ObservationHelper.java:52]: Or alternatively, could the methods be static and it not have a public constructor?

Comment (johkin) [reply spring-ws-core/src/main/java/org/springframework/ws/support/ObservationHelper.java:52]: It could certainly be a singleton, is something like this what you had in mind:
`java

   private static ObservationHelper INSTANCE;

    public static ObservationHelper getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new ObservationHelper();
        }
        return INSTANCE;
    }

    private ObservationHelper() {
`

Also noted that a minor bug slipped through the tests, I have a fix in place locally.

Comment (shakuzen) [reply spring-ws-core/src/main/java/org/springframework/ws/support/ObservationHelper.java:52]: Yes, that would be one option. I'll leave it up to @corneil to advise on what makes most sense given existing patterns used in this codebase.

**Comment by corneil** [comment]
@johkin when I checkout the PR and run `./mvnw clean package` there are test failures:

```
[ERROR] Failures: 
[ERROR]   WebServiceTemplateObservationIntegrationTest.attachment:305 There should be at least one Observation that matches the assertion. Found following Observations:
name='webservice.client', contextualName='POST', error='null', lowCardinalityKeyValues=[exception='none', host='localhost', localpart='unknow', namespace='unknown', outcome='success', soapaction='none'], highCardinalityKeyValues=[path='/soap/attachment'], map=[], parentObservation=null
[ERROR]   WebServiceTemplateObservationIntegrationTest.marshalSendAndReceiveNoResponse:222 There should be at least one Observation that matches the assertion. Found following Observations:
name='webservice.client', contextualName='POST', error='null', lowCardinalityKeyValues=[exception='none', host='localhost', localpart='unknow', namespace='unknown', outcome='success', soapaction='none'], highCardinalityKeyValues=[path='/soap/noResponse'], map=[], parentObservation=null
[ERROR]   WebServiceTemplateObservationIntegrationTest.notFound:241 There should be at least one Observation that matches the assertion. Found following Observations:
name='webservice.client', contextualName='POST', error='org.springframework.ws.client.WebServiceTransportException: Server Error [500]', lowCardinalityKeyValues=[exception='WebServiceTransportException', host='localhost', localpart='unknow', namespace='unknown', outcome='fault', soapaction='none'], highCardinalityKeyValues=[path='/errors/notfound'], map=[], parentObservation=null
[ERROR]   WebServiceTemplateObservationIntegrationTest.receiverFault:262 There should be at least one Observation that matches the assertion. Found following Observations:
name='webservice.client', contextualName='POST', error='org.springframework.ws.soap.client.SoapFaultClientException: Receiver Fault', lowCardinalityKeyValues=[exception='SoapFaultClientException', host='localhost', localpart='unknow', namespace='unknown', outcome='fault', soapaction='none'], highCardinalityKeyValues=[path='/soap/receiverFault'], map=[], parentObservation=null
[ERROR]   WebServiceTemplateObservationIntegrationTest.sendSourceAndReceiveToResult:158 There should be at least one Observation that matches the assertion. Found following Observations:
name='webservice.client', contextualName='POST', error='null', lowCardinalityKeyValues=[exception='none', host='localhost', localpart='unknow', namespace='unknown', outcome='success', soapaction='none'], highCardinalityKeyValues=[path='/soap/echo'], map=[], parentObservation=null
[ERROR]   WebServiceTemplateObservationIntegrationTest.sendSourceAndReceiveToResultNoResponse:177 There should be at least one Observation that matches the assertion. Found following Observations:
name='webservice.client', contextualName='POST', error='null', lowCardinalityKeyValues=[exception='none', host='localhost', localpart='unknow', namespace='unknown', outcome='success', soapaction='none'], highCardinalityKeyValues=[path='/soap/noResponse'], map=[], parentObservation=null
[ERROR]   WebServiceTemplateObservationIntegrationTest.senderFault:282 There should be at least one Observation that matches the assertion. Found following Observations:
name='webservice.client', contextualName='POST', error='org.springframework.ws.soap.client.SoapFaultClientException: Sender Fault', lowCardinalityKeyValues=[exception='SoapFaultClientException', host='localhost', localpart='unknow', namespace='unknown', outcome='fault', soapaction='none'], highCardinalityKeyValues=[path='/soap/senderFault'], map=[], parentObservation=null
[ERROR]   ObservationInterceptorIntegrationTest.testObservationInterceptorBehavior:102 There should be at least one Observation that matches the assertion. Found following Observations:
name='webservice.server', contextualName='POST /ws/{pathInfo}', error='null', lowCardinalityKeyValues=[exception='none', localpart='unknow', namespace='unknown', outcome='success', path='/ws/{pathInfo}', soapaction='none'], highCardinalityKeyValues=[pathinfo='/1234'], map=[], parentObservation=null,name='webservice.server', contextualName='POST /ws', error='null', lowCardinalityKeyValues=[exception='none', localpart='unknow', namespace='unknown', outcome='success', path='/ws', soapaction='none'], highCardinalityKeyValues=[], map=[], parentObservation=null
[ERROR]   ObservationInterceptorIntegrationTest.testPathWithVariable:125 There should be at least one Observation that matches the assertion. Found following Observations:
name='webservice.server', contextualName='POST /ws/{pathInfo}', error='null', lowCardinalityKeyValues=[exception='none', localpart='unknow', namespace='unknown', outcome='success', path='/ws/{pathInfo}', soapaction='none'], highCardinalityKeyValues=[pathinfo='/1234'], map=[], parentObservation=null
```

**Comment by johkin** [comment]
@corneil Did you see my fix for the tests, and also if you have any preference for the solution of ObservationHelper? Should it be a singleton in code or a singleton managed by Spring?

**Review by corneil — APPROVED** [review]
This looks good. 
Thank you @johkin 

I will merge when the build is stable.

**Comment by johkin** [comment]
@snicoll 
Any chance that this would make it into a M2 milestone?

**Comment by snicoll** [comment]
I honestly don’t know.

I’ve started to review the initial commit with @bclozel and it needs more work. We’re also late in the 4.1 cycle with GA happening in May already.

**Comment by snicoll** [comment]
@johkin we've looked at it and we've started 4.1.x dev very late for the May train so this will have to move.

**Comment by johkin** [comment]
If there are anything that could be done to help out, please let me know.

**Pull request merged**
